src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion.py

```diff
@@ -0,0 +1,312 @@
+
+from ...models import AutoencoderKL, UNet2DConditionModel, VQModel
+from ...models.attention import Transformer2DModel
+from ...pipeline_utils import DiffusionPipeline, ImagePipelineOutput
+from ...schedulers import DDIMScheduler, LMSDiscreteScheduler, PNDMScheduler
+
+
+class VersatileMixedModel:
+    """
+    A context managet that swaps the transformer modules between the image and text unet during inference, depending on
+    the latent type and condition type.
+    """
+
+    def __init__(self, image_unet, text_unet, latent_type, condition_type):
+        self.image_unet = image_unet
+        self.text_unet = text_unet
+        self.latent_type = latent_type
+        self.condition_type = condition_type
+
+    def swap_transformer_modules(self):
+        for name, module in self.image_unet.named_modules():
+            if isinstance(module, Transformer2DModel):
+                parent_name, index = name.rsplit(".", 1)
+                index = int(index)
+                self.image_unet.get_submodule(parent_name)[index], self.text_unet.get_submodule(parent_name)[index] = (
+                    self.text_unet.get_submodule(parent_name)[index],
+                    self.image_unet.get_submodule(parent_name)[index],
+                )
+
+    def __enter__(self):
+        if self.latent_type != self.condition_type:
+            self.swap_transformer_modules()
+        return self.image_unet if self.latent_type == "image" else self.text_unet
+
+    def __exit__(self, *exc):
+        # swap the modules back
+        if self.latent_type != self.condition_type:
+            self.swap_transformer_modules()
+
+
+class VersatileDiffusionPipeline(DiffusionPipeline):
+    r"""
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+
+    Parameters:
+        vqvae ([`VQModel`]):
+            Vector-quantized (VQ) Model to encode and decode images to and from latent representations.
+        bert ([`LDMBertModel`]):
+            Text-encoder model based on [BERT](https://huggingface.co/docs/transformers/model_doc/bert) architecture.
+        tokenizer (`transformers.BertTokenizer`):
+            Tokenizer of class
+            [BertTokenizer](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertTokenizer).
+        unet ([`UNet2DConditionModel`]): Conditional U-Net architecture to denoise the encoded image latents.
+        scheduler ([`SchedulerMixin`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents. Can be one of
+            [`DDIMScheduler`], [`LMSDiscreteScheduler`], or [`PNDMScheduler`].
+    """
+    tokenizer: CLIPTokenizer
+    image_processor: CLIPProcessor
+    text_encoder: CLIPTextModel
+    image_encoder: CLIPVisionModel
+    image_unet: UNet2DConditionModel
```

Comment:
Do we really need two unets here? I think we should only have one unet -> we have one denoising loop we should use just one unet for this no?

Reply:
For text2img we need the body of `image_unet` and transformer2d blocks from `text_unet` to swap them out

Reply:
only img2img and text2text (AKA variations) use one unet